client/src/pages/DriverLocationInfo/AssignLocationDropdown.tsx

```diff
@@ -69,64 +58,115 @@ const AssignLocationDropdownItem: React.FC<DropdownItemProps> = (
+  idx: number,
+  field: keyof DropoffLocationUIState,
+  newState: boolean
+) {
+  const newLocationsUIStateList = [...prevUIStateList];
+  newLocationsUIStateList[idx][field] = newState;
+  return newLocationsUIStateList;
 }
 
-//TODO: Error handling on assigning a dropoff location doesn't work
-export const AssignLocationDropdown: React.FC<Props> = ({
-  locations,
-  driver,
-  refetchDrivers,
-}) => {
+export const AssignLocationDropdown: React.FC<{
+  locations: ProcessedDropoffLocation[];
+  driver: ProcessedDriver;
+}> = ({ locations, driver }) => {
   const { token } = useAuth();
   if (!token) {
     return <Navigate to="/" />;
   }
-  const [selectedLocations, setSelectedLocations] = useState<boolean[]>(
-    getSelectedLocations(locations, driver)
-  );
+
+  const queryClient = useQueryClient();
+  const [dropoffLocationsUIStateList, setDropoffLocationsUIStateList] =
+    useState<DropoffLocationUIState[]>(
+      getDropoffLocationsUIStateList(locations, driver)
+    );
   const [isDropdownOpen, setIsDropdownOpen] = useState(false);
 
-  const assignLocation = (id: string, idx: number) => async () => {
-    const locationIds = selectedLocations[idx] ? [] : [id];
-    for (let i = 0; i < selectedLocations.length; i++) {
-      if (i != idx && selectedLocations[i]) {
-        locationIds.push(locations[i].id);
+  const assignLocationMutation = useMutation({
+    mutationFn: async ({ id, idx }: { id: string; idx: number }) => {
+      //Payload to update this driver's locations has to contain all of the selected locations for this driver.
+      //If the location was previously selected, then that means we are unassigning that location to this driver.
+      //If the location was NOT previously selected, then that means we are assigning that location to this driver.
+      const locationIds = dropoffLocationsUIStateList[idx].isSelected
+        ? []
+        : [id];
+      //Add the rest of the locations for this driver to the payload
+      for (let i = 0; i < dropoffLocationsUIStateList.length; i++) {
+        if (i != idx && dropoffLocationsUIStateList[i].isSelected) {
+          locationIds.push(locations[i].id);
+        }
       }
-    }
-    const url = `/api/volunteers/drivers/assign-location/${driver.id}`;
-    const resp = await fetch(url, {
-      method: "PATCH",
-      headers: {
-        Authorization: `Bearer ${token}`,
-        "Content-Type": "application/json",
-      },
-      body: JSON.stringify({ locationIds: locationIds }),
-    });
-    if (!resp.ok) {
-      const data = await resp.json();
-      throw new Error(data.message);
-    }
-    return resp.json();
-  };
+      //Set the state for the newly selected/unselected location to loading
+      setDropoffLocationsUIStateList((prev) =>
+        updateDropoffLocationsUIStateList(prev, idx, "isLoading", true)
+      );
+      const url = `/api/volunteers/drivers/assign-location/${driver.id}`;
+      await fetch(url, {
```

Comment:
The onError in react-query on gets called if an error is thrown in the call. fetch does not throw if the status != 200 like other libs such as axios. 

We should do 
```ts
if (!response.ok) {
  throw Error("Failed in fetch");
}
```

for the onError to work.

Please let me know if I am correct.

Reply:
Good catch. Your suggestion matches the use in this [file](https://github.com/grassrootsgrocery/admin-portal/blob/e54c100d227d2a7615b33a2e4d66784dde10561c/client/src/utils/http.ts).

Reply:
@jasoncavanaugh, once this change is in I think we can go ahead and merge. I played around with the deployed PR and everything looks to be working fine. I'll mark as approved for now.